Recover file work over

zonkmachine opened the pull request:
Fixes issue: #2174
Edit: I've started a feedback thread in the forum https://lmms.io/forum/viewtopic.php?f=15&t=3479

In summary: Multiple fixes for the recovery file system that introduces clear session states for LMMS.
It won't allow overwriting of ```recover.mmp``` if you run multiple instances but it will only do autosave for the first instance of LMMS launched. 

![recovery-final](https://cloud.githubusercontent.com/assets/6368949/11822709/3b6f9830-a370-11e5-832a-6fee16945a96.png)

---

Bugs and fixes!
 - If you start LMMS with a recovery file present and don't do anything at all the auto save will write it over with an all empty project after a minute. It's been ticking on in the background. We solve this by starting the timer the very last thing in main() after a project is loaded( if it's not a limited session or auto save is disabled altogether ).

 - When starting if you escape from the recover message the default is to delete recover.mmp  This is probably not expected or wanted behaviour for everyone. Escaping from the dialogue had better be non destructive.
I fix this by defaulting 'Escape' to button 'Exit' instead close the app.

 - If you press yes to recover and immediately shut down LMMS you will not be asked any questions and recover.mmp is deleted. This is a bit rough. Fixed in the new file save dialogue. ( MainWindow::mayChangeProject() message box )

 - If you run multiple sessions of LMMS they will simply take turns to write over recover.mmp  :P
I solve this by introducing the limited session, that will be set if the user press 'Ignore' at the recovery dialogue.  This allows you to launch another session that doesn't write to recover.mmp. This way LMMS will launch a slightly limited session which simply turns off autosave for this instance. This together with running autosave() the first thing of all (described below) should minimise the risk of recover.mmp being overwritten.

 - On starting LMMS the autosave() function takes a minute to kick in. This is problematic especially since the introduction of different sessions, because this minute of no backup file is actually a logic state of it's own, and it's an ill defined one. We take care of this by executing ```MainWindow::autosave()``` in main() once as a very last step after loading a project, but only if there is no recover file present. For this to work I made MainWindow::autosave() public.

 - Clarity. Brushed up the dialogues involved and update window title to signal session.

I know it's not 100% what users have asked for but it is a solid solution for recover files which has no uncertain states and is hopefully clear enough to use. It's intended as a modification/bugfix for the upcoming 1.2 release.
Separate files for multiple sessions as has been in demand will most likely have to wait until later.
The actual action of the MainWindow::autosave() when it's running however will be one of my next areas to look at.

Last updated on 15 Dec 2015

zonkmachine commented:
Separate messages for the cases where you have a modified unsaved file or a recovered file. If you modify a recovered file it will return the recovered file message.

zonkmachine commented:
In the last commit I create an empty file RECOVER in the same directory as the recover.mmp file and check for it's existance instead of the latter to avoid the recover files created during operation.
Now the logic seems to work fine.

zonkmachine commented:
Launching subsequent instances of LMMS should ignore ongoing recover session and not touch recover.mmp and RECOVER.

zonkmachine commented:
undefined reference to `MainWindow::recoverSessionCleanup()' ... ?

Yeah, I didn't think the last one would pass. I need some feedback here.
Everything was rolling along just fine until I got the mighty fine idea to put some commands in a simple function. these QFile::remove commands works fine in the MainWindow::closeEvent by themselves but not when in the function recoverSessionCleanup()

```c++
void recoverSessionCleanup()
{
	// delete recover session files
	QFile::remove( ConfigManager::inst()->recoveryFile() );
	QFile::remove( ConfigManager::inst()->recoveryFile() + ".bak" );
	QFile::remove( ConfigManager::inst()->recoverSessionFile() );
}
```


```c++
void MainWindow::closeEvent( QCloseEvent * _ce )
{
	if( mayChangeProject(true) )
	{
		if( isRecoverSession() )
		{
			recoverSessionCleanup();
		}
		_ce->accept();
	}
	else
	{
		_ce->ignore();
	}
}
```

Travis says:
```
Linking CXX executable tests

../src/CMakeFiles/lmmsobjs.dir/gui/MainWindow.cpp.o: In function `MainWindow::closeEvent(QCloseEvent*)':

/home/travis/build/LMMS/lmms/src/gui/MainWindow.cpp:1283: undefined reference to `MainWindow::recoverSessionCleanup()'

collect2: ld returned 1 exit status

make[3]: *** [tests/tests] Error 1

make[2]: *** [tests/CMakeFiles/tests.dir/all] Error 2

make[1]: *** [tests/CMakeFiles/tests.dir/rule] Error 2

make: *** [tests] Error 2
```

zonkmachine commented:
I squashed it down to two commits and reverted to the last working state by commenting out the function above, recoverSessionCleanup() .
Changes so far are:
 * Forces question on exit with recover.mmp
Before you could just open and close the file with no question asked and the files would be deleted on exit.
 * Only one instance of LMMS can access recover.mmp
 * Separate question on close for a modified file and a recovered, for clarity.

---
What doesn't work?

If you open a project or create a new default one during a recover session, which forces the right message box, you will now remain in a recoverSession and be prompted as such at the next message box. The program will also not clear up any of the files involved.

Wallacoloo commented:
By the way, it looks like all you need to do is change
```c++
void recoverSessionCleanup()
{
[...]
}
```
to
```c++
void MainWindow::recoverSessionCleanup()
{
[...]
}
```

zonkmachine commented:
You mean like ALL the other functions...  :-P

zonkmachine commented:
@Wallacoloo Thanks! Now I have a working patch for the recover file. It turned out to need a little bit more than the 5 - 10 lines I was hoping for but there are many places to exit an ongoing project and you need to cover all those events.

What I've done is create a file RECOVER which lets other instances know that they don't need to care about the recover.mmp file. There is also a bool m_recoverSession which we set to make LMMS aware of that it is doing recover work. We can check session state whith 'MainWindow::isRecoverSession()'

 * Forces question on exit with recover.mmp Before you could just open and close the file with no question asked and the files would be deleted on exit. This works if you 'exit the recover session' <by Ctrl + N>, <Ctrl + q>, <Ctrl + o> or with equivalent menu action.
 * Only one instance of LMMS can access recover.mmp
 * Separate question on close for a modified file and a recovered, for clarity.
 * Sett window title to further distinguish session from a new default project ( next commit )

---

Issues:
 * If a recover session crash, or is terminated from the command line, RECOVER remains and needs to be manually removed. Maybe a message box on start to check if there really is an ongoing recover session that can then remove the files. I'll work on this next.
 * Has the potential to be annoying...

zonkmachine commented:
Like this the user can recover from a locked recover session.

![recovermessagebox2](https://cloud.githubusercontent.com/assets/6368949/8640615/be2dd3d2-28fe-11e5-9112-6bdb89f15e5b.png)

zonkmachine commented:
This is now a working fix for #2174 and is best described here https://github.com/LMMS/lmms/pull/2176#issuecomment-120720126
I can't force it into any unwanted states and it takes good care of recover.mmp
If it's annoying to no ends? Don't know, works for me for now.

tresf commented:
"Open" might be misleading.... As an example, here's what VIM says:

```
E325: ATTENTION
Found a swap file by the name ".test.txt.swp"
          owned by: tres   dated: Sun Jul 12 20:26:57 2015
         file name: ~tres/test.txt
          modified: no
         user name: tres   host name: ubuntu-1204
        process ID: 5934 (still running)
While opening file "test.txt"
             dated: Wed May 14 23:18:12 2014

(1) Another program may be editing the same file.  If this is the case,
    be careful not to end up with two different instances of the same
    file when making changes.  Quit, or continue with caution.
(2) An edit session for this file crashed.
    If this is the case, use ":recover" or "vim -r test.txt"
    to recover the changes (see ":help recovery").
    If you did this already, delete the swap file ".test.txt.swp"
    to avoid this message.

Swap file ".test.txt.swp" already exists!
[O]pen Read-Only, (E)dit anyway, (R)ecover, (Q)uit, (A)bort:
```

-Tres

tresf commented:
BTW, I'm not a fan of the `RECOVER` file.  I think we should refrain from using a fixed file name and instead keep our current implementation and fix it rather than adding even more recovery temp files.

zonkmachine commented:
> "Open" might be misleading.... As an example, here's what VIM says:

I scavenged among the default buttons. I haven't looked into adding custom buttons to QMessageBox yet,
I wanted button names, Recover, Discard and... well, something else than ignore.



> BTW, I'm not a fan of the RECOVER file. I think we should refrain from using a fixed file name and instead keep our current implementation and fix it rather than adding even more recovery temp files.

OK. I'm not to happy about the extra file either but it kind of work. I can just remove it and have a more general message like the one from Wim about the dire consequences of running multiple instances.

zonkmachine commented:
There is a one line fix for this that does exactly what the title proposes.
add Engine::getSong()->setModified(); after loading recover.mmp in main.

This would mean you wouldn't be able to just open and close recover.mmp and it's gone, which was what triggered the bug report in the first place. We could do this for 1.2 and do a more complete fix later. Being in a freeze and all. I think however it's a good idea to really have separate messages.

zonkmachine commented:
Simplified. File RECOVER is gone and the new Messagebox.

---

Changes in summary:
 * Forces question on exit with recover.mmp Before you could just open and close the file with no question asked and the files would be deleted on exit. This works if you 'exit the recover session' , , or with equivalent menu action.
 * Separate question on close for a modified file and a recovered, for clarity.
 * Sett window title to further distinguish session from a new default project ( next commit )

zonkmachine commented on src/gui/MainWindow.cpp:
```diff
@@ -644,17 +649,31 @@ void MainWindow::resetWindowTitle()
+					"unsaved and will be lost if you don't "
+					"Save it. Do you want to save it now?" );
+
+	QString messageTitleUnsaved = tr( "Project not saved" );
```
What happens with the already translated string below when I move it to a variable? I didn't change the text. Do I need to do anything to link it to the old string?

midi-pascal replied:
Since the strings are translated at creation time by using tr() it should be Ok. You can then copy to an other variable because the content is already translated.
So this should work as expected :)

zonkmachine replied:
Cool, thanks!  8)

zonkmachine commented:
![recoverqoriginal](https://cloud.githubusercontent.com/assets/6368949/8750427/8a0ed022-2ca7-11e5-8a23-b2fa91a266e8.png)

Here's the messagebox in master. I've struck a somewhat odd problem. If you try and close the messagebox with the [x] your recover.mmp is discarded. In a basic messagebox the closeEvent() isn't defined so you have to take care of this yourself but in the simpler version of messagebox with default buttons closeEvent() takes care of closing the window just fine but also seem to return the enum of the button with the highest value. OK is lowest...:
http://doc.qt.io/qt-4.8/qmessagebox.html#StandardButton-enum
I haven't come up with a good way to deal with this yet.
What's the response from closeEvent() supposed to be in our case?

Apart from that the only other thing I can think of improving on for this pull request is the message itself.

zonkmachine commented:
Got it!
```If there is exactly one button having either the Reject role or the the No role, it is the button activated when Esc is pressed.```

zonkmachine commented:
```[O]pen Read-Only, (E)dit anyway, (R)ecover, (Q)uit, (A)bort:```
What's the difference of quit and abort in this context. One closes the program and the other opens a default empty project?

![messagebox2 3](https://cloud.githubusercontent.com/assets/6368949/8759531/35042e7a-2cf6-11e5-8a3d-de49d5c5ca7e.png)

Buttons renamed. I changed 'escape/close' button to a new 'close' button instead of as today, 'no' (discard).
In this sense it works pretty much as it did before but has unwanted responses fixed and a bit more text. Old message is reused however and if the translation isn't updated it should still be pretty clear.

tresf commented:
"Close" may be a but confusing, as it is often associated with the closing of a dialog...

zonkmachine commented:
Quit or Abort then?

tresf commented:
Yeah... I think.  Quit, or Quit LMMS.

Quit LMMS seems a bit overkill, but the action is clear.  We could use a wildcard to replace LMMS with application name too, if desired.

zonkmachine commented:
I prefer Quit over Quit LMMS then.
> We could use a wildcard to replace LMMS with application name too, if desired.

You lost me there. What other app?

tresf commented:
Oh, I just meant the global name "LMMS" to prevent hard-coding.

zonkmachine commented:
![messagebox2 5](https://cloud.githubusercontent.com/assets/6368949/8759884/1fae04b4-2d01-11e5-8140-f9bd841ab8a7.png)

zonkmachine commented:
![messagebox2 6](https://cloud.githubusercontent.com/assets/6368949/8763470/d574ff4a-2d9d-11e5-9adc-6db3f09e09d2.png)

I've reimplemented the possibility to open a limited session without autosave that was lost with the file RECOVER. I'm not happy with the name 'Open' for the button though.
There is one more thing now that's on my mind. autosave() hits in first after a minute and during this time plenty can happen. Like launching a second instance of LMMS. One way to go about this is to trigger autosave() once after setting up the session in main.cpp . I've tried this and it works just fine but I had to make autosave() public. I'm currently looking through the Qt manual to see if it's possible to run the timer much like  a do-while loop. Execute immediately first run and then timer on.

Umcaruje commented:
>I'm not happy with the name 'Open' for the button though.

I think 'Ignore' would be a good name for that button.

zonkmachine commented:
> I think 'Ignore' would be a good name for that button.

That's better yes.

zonkmachine commented:
![messagebox2 7](https://cloud.githubusercontent.com/assets/6368949/8763657/4bb9b72c-2da8-11e5-885b-ba7f0d7e7be6.png)


Yeah, 'Ignore' has the right touch to it. 8)

> I'm currently looking through the Qt manual to see if it's possible to run the timer much like a do-while loop.

A no go. I pushed the first version instead.

zonkmachine commented:
Closing the messagebox ( Escape / [x] / Alt+F4 ) right now defaults to 'Close'. With the new button perhaps Ignore should be the default response?

musikBear commented:
now this is already addressed, and i should properly keep my big trap shut, but what is wrong with a concatenated name for the recover-file
The recover of 'myProject.mmpz' becomes 'myProjectRecover.mmp'
Then overwrite and all that is not an issue at all..
(a need for temporary manual cleanup by user is, and i understand that this would be a weighted decision

zonkmachine commented:
I'm first shaping up the existing code. Then I'm going to look into what new functions to take in. Remember that we are working toward the 1.2 release and are in a freeze so what I'm basically doing is fixing everything that is a bug in the context of the original code.

zonkmachine commented:
![recovermenu3 0](https://cloud.githubusercontent.com/assets/6368949/8765921/bde1b062-2e24-11e5-96c1-a17de09ce2f2.png)

I fear my home brew layout with tabs and line breaks isn't going to work well with translations so I went shopping for something better. As it turns out qmessagebox knows HTML... :P
I don't know if qlinguist can manage one chunk of HTML thrown at it like this though.

tresf commented:
> I don't know if qlinguist can manage one chunk of HTML thrown at it like this though.

Perhaps just translate the cells and hard-code the layout?

zonkmachine commented:
Looks like Qt should handle it though. From here:
http://stackoverflow.com/questions/16664915/how-to-use-html-inside-of-translatable-strings-in-qt

```
The ts file is XML. If you want to use HTML tags you will
need to escape them. For example: <source>&lt;b&gt;text&lt;/b&gt;</source>

Note that the recommended way of managing translations
in Qt is to run the lupdate tool on your source code and have
Qt generate the .ts file, in which case this will be done automatically.
```

tresf commented:
I would not recommend putting any npn-essential HTML in the translation file.  The HTML can be static.

Umcaruje commented:
>Closing the messagebox ( Escape / [x] / Alt+F4 ) right now defaults to 'Quit'. With the new button perhaps Ignore should be the default response?

I think that Quit is better than Ignore, since some users might expect that closing the dialog would function as either 'Discard' or 'Ignore', and there is no way to differentiate from those two without looking for 'recover.mmp' in your lmms folder. If the behaviour is 'Quit', its pretty straightforward and the user has to choose.
***
One minor thing: I think you should either have icons on all buttons, or have no icons at all. Right now it doesn't look very consistent.

grejppi commented:
> This is not reversible!

Compare:
> This is not reversible.

Which would you prefer, a dialog that just states the facts, or one that yells at the user's face?

tresf commented:
How about 

> This is **<u>NOT</u>** reversible!!!!

zonkmachine commented:
> I would not recommend putting any nonessential HTML in the translation file. The HTML can be static.

Nothing fancy. Just header, bold and table tags and a star trek face palm meme as pop-up when you choose the wrong button.

> One minor thing: I think you should either have icons on all buttons, or have no icons at all. Right now it doesn't look very consistent.

I will comply.

> Which would you prefer, a dialog that just states the facts, or one that yells at the user's face?

They're both totally awesome! You pick!

> How about
> 
>     This is NOT reversible!!!!

Totally awesome!

zonkmachine commented:
I updated the first post to sum up the work so far.

Spekular commented:
There seems to be some inconsistency in what the recover file is called. We
have "recover *the file*", "the *recover file*" and "the *restored files*".
I think "recovered" is more accurate than "restored", and singular is
probably better than plural. That said I don't like how "the recover file"
sounds (to me), so I'd suggest "recovered file"/"recovery file".

To adress the naming of the file, I don't think anyone has suggested we do
it the way blender does, by creating fileName.blend1 (and sometimes
.blend2, maybe so if it crashes while overwriting one backup you still have
the other). I think maybe fileName.mmpz.bak or some similar added extension
would be better, but maybe that's just me.

zonkmachine commented:
> There seems to be some inconsistency in what the recover file is called. We have "recover *the file*", "the *recover file*" and "the *restored files*". I think "recovered" is more accurate than "restored", and singular is probably better than plural. That said I don't like how "the recover file" sounds (to me), so I'd suggest "recovered file"/"recovery file".

Thanks! I've renamed the stray 'restored', changed to singular, settled for 'recovery file'.
I'll look into naming the recover file after an existing project name.

tresf commented:
Renaming the recover file can make this task tougher.   It would likely re-introduce the double-recovery file design unless we have an "untitled.mmpz" in the lmms home for unnamed files, and meanwhile each file checked for a dopleganger before opening.  I too believe this to be ideal, but might be out if the scope of this PR.

tresf commented:
P.S. @zonkmachine, tremendous work on the reception to feedback here.  Top notch. :+1:

zonkmachine commented:
@tresf Thanks!

> Renaming the recover file can make this task tougher.

Tried some quick hacks and backed off. It will be for another day.

> I too believe this to be ideal, but might be out if the scope of this PR.

This! I see new issues pop up all the time and need to keep focus. I'd prefer to get something merged first before expanding the work. I'm also thinking of waiting with the HTML work until later. I like the look of it but there are too many question marks there right now.
Also, there are other things to fix and I'm currently locked up here.

zonkmachine commented:
New bug. Probably my favorite so far.
In master try this:
1. Launch LMMS with a recover.mmp present ( one that you don't mind to lose... )
2. Wait.

That's it. You don't need to do anything at all. After one minute autosave() is launched without main() having launched a proper project so recover.mmp is overwritten with an empty project with no tracks.

Spekular commented:
I'd just like to second tresf's sentiment, it's incredible how quickly
you've responded to comments! :+1:

zonkmachine commented:
@Spekular Thanks!

---

I'm now trying to remove the limited session to simplify the code. Basically what it means anyway is to load a normal LMMS session but without the file deletion action that happens on exit. We can simply remove this from a normal session and just load a normal session on 'ignore'. You still have the option to delete files on launch by hitting 'Discard' and this should be enough.
I'll try and simplify the code as much as possible now and see what bugs there is left. There hasn't been mush compile problems on this project but on the other hand a major juggling of logic states so my brain is in a somewhat bruised state right now.

zonkmachine commented:
> I'm now trying to remove the limited session to simplify the code

Fail! That didn't work. I also tried to squash the commits down so they more clearly reflected the work done but so far that hasn't been successful. Furthermore I tried to produce a graf of what's been done here but libreoffice snap to grid function won after an uneven battle for some two hours and what I've got now is ascii art instead. But only for main.cpp so far.

Anyway, I think it's working now as I intended it. Basically the 'Ignore' function makes out some ~80% of whatever pain that is involved in this pull request. If we would just fix the recovery function up in the most straightest way that would mean forcing the user to deal with the recovery file straight away and that would probably be annoying. Now you can press 'Ignore if you for instance want to run another session simultaneously or whatever your heart tells you. I just ran three sessions at once, playing and editing, and only the first wrote to recover.mmp . It needs testing off course and still has code commented out left in.

Off course, all this is, strictly speaking, a matter of form as there will be no crashes in the final 1.2 release.

zonkmachine commented on include/MainWindow.h:
```diff
@@ -30,6 +30,7 @@
 #include <QtCore/QList>
 #include <QMainWindow>
 
+#include "ConfigManager.h"
```
I don't remember why this moved... I think it was when I had problems compiling ( earlier in this thread ).
Moved back.

zonkmachine commented on include/MainWindow.h:
```diff
@@ -149,6 +187,8 @@ public slots:
 	void refocus();
 
+	bool m_recoverSession;
+	bool m_limitedSession;
```
I started out with recoverSession and then added 'limited' along the way. Maybe this should be done with an enum instead, something like:
enum Session { Normal, Recover, Limited }; ?

musikBear commented:
Sadly no one responds to your forum post. You know my feeling about 'recover', so i have not felt it necessary to post  - i have made it 'sticky' though, so it should stay in focus of all visitors
Should it be moved to *'Announcements'*

tresf commented:
> Should it be moved to 'Announcements'

I'll let @zonkmachine decide, but he's had quite a bit of valuable feedback here, the forum post can probably be taken off of sticky if.

zonkmachine commented:
@musikBear Thanks, sticky isn't needed.

zonkmachine commented:
> > I don't know if qlinguist can manage one chunk of HTML thrown at it like this though.
> 
> Perhaps just translate the cells and hard-code the layout?

It looks like that message box isn't translated anyway so right now it isn't a big thing anyway. I'd just go with either of them. The HTML version looks better but I'm pretty happy with the way it looks now. One sneaky way to do it is to pop the HTML version in there an lift out the tr() stuff since no strings are translated anyway, so we can go back and change that later when we know how it works.

zonkmachine commented:
Here are the basic layout of the changes in main() .

![lmms_main cpp_before](https://cloud.githubusercontent.com/assets/6368949/8876247/d6d1a002-321e-11e5-9edb-86711644e13d.png)

![lmms_main cpp_after](https://cloud.githubusercontent.com/assets/6368949/8876257/dc686532-321e-11e5-9b37-82a2ca15f40a.png)

tresf commented:
> It looks like that message box isn't translated anyway so right now it isn't a big thing anyway. I'd just go with either of them. The HTML version looks better but I'm pretty happy with the way it looks now.

Well, that's not a great reason to leave it as-is.  Any translation of this dialog has the potential to break formatting, and it is not fair to put that on the translators, is it?

If HTML displays properly, I would propose something like this (pseudocode).


```cpp
QString("
    <tr><td>%1</td><td>%2</td></tr>
    <tr><td>%3</td><td>%4</td></tr>
    </html>
").arg(tr("Foo"), tr("Bar"), tr("Fizz"), tr("Buzz");
```

zonkmachine commented:
OK. I'll have another look at the HTML version.

zonkmachine commented:
> If HTML displays properly, I would propose something like this (pseudocode).
> ```c++
QString("
    <html><table>
        <tr><td>%1</td><td>%2</td></tr>
        <tr><td>%3</td><td>%4</td></tr>
    </table></html>
").arg(
    tr("Foo"), tr("Bar"),
    tr("Fizz"), tr("Buzz")
);
```

That was clear instructions, thanks!
HTML pushed with translation strings.

tresf commented:
This looks nice.  I feel like we may have a redundant MainWindow::tr( ) around the whole html text though.  Am I seeing that right?

zonkmachine commented:
> a redundant MainWindow::tr( )

Ah! From the first translation efforts.

zonkmachine commented:
Fixed!

zonkmachine commented:
Like the last commit. A bit better I think.
if it's a limited session the timer isn't started in main() and when we run occasional autoSave() to counter for a changing of project or starting a new one, etc.  we do it with runAutoSave() instead which filters out when it's a limited session or when auto save isn't enabled at all.

zonkmachine commented on src/core/main.cpp:
```diff
@@ -451,14 +452,87 @@ int main( int argc, char * * argv )
+			}
+			else // ::Abort
+			{
+				return 0;
```
Do I need to go 'delete app;' on this one before returning?

zonkmachine commented:
I'm testing it with multiple instances now under gdb.
I don't think this backtrace is from anything I did...

```gdb
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffceca1700 (LWP 13720)]
0x00000000004bd4d1 in FxMixer::masterMix(float (*) [2]) ()
(gdb) bt
#0  0x00000000004bd4d1 in FxMixer::masterMix(float (*) [2]) ()
#1  0x00000000004dc6b2 in Mixer::renderNextBuffer() ()
#2  0x00000000004dc974 in Mixer::fifoWriter::run() ()
#3  0x00007ffff685a32f in ?? () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#4  0x00007ffff7bc4182 in start_thread (arg=0x7fffceca1700) at pthread_create.c:312
#5  0x00007ffff41dd47d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb)
```

midi-pascal commented:
@zonkmachine For sure your code is not the cause of this SIGSEGV. I think you hit  issue #2192 which is a really nasty one since it happens randomly.

zonkmachine commented:
@midi-pascal Thanks! That looks like the same case.

---

Unless you do something silly now like:
 * saving your files in the working directory...
 * while taking advance money for jobs you do...
 * ant trusting on the recovery function to also be your only backup system...
 * and you get to remix your favourite artist https://www.youtube.com/watch?v=MRiFMLSq13U ...!

I see how that could be a problematic case on many levels. :)

I'm running out of cases to test but I'm going to try and load defective files and other odd stuff to see what happens.

zonkmachine commented:
Some cases that I set as goal for the project and that now works fine.

Test case 1:

 1 -  Open a known track in LMMS
 2  - Close it from the terminal thus creating a reference recovery file
 3 - Open a couple of separate sessions and choose 'Ignore'. From within these session just do whatever related to opening and saving files ( but don't open the reference project! ).
 4 - Close down all instances of LMMS
 5 - Start LMMS again, choose recover and confirm that the file is the same as the reference project.

Test case 2:

 1 - Repeat 1 and 2 from the case above.
 2 - Start LMMS and in settings, switch 'Enable auto save feature' off.
 3 - Close LMMS
 4 - Repeat step 3 from test case 1 above and observe that you will be presented the same recovery menu as previously, but now when you choose ignore you are in fact working in a normal LMMS session with auto save off. No reminders to save in the title bar, just the usual message box on exiting your work.
 5 - Close all instances of LMMS and confirm that you get the reference project when you re-run LMMS and choose 'Recover'.

zonkmachine commented:
Fixed a bug when loading with a file as argument.

---

```bash
Session cleanup
Session cleanup
```

Sometimes cleanup is done twice. This keeps us from having to do unnecessary checks in a whole lot of places. Probably. I don't quite remember what the issue was here but I'll have another go at fixing it.
It won't be noticed anyway because those debug messages aren't staying for long.

zonkmachine commented:
Text and layout. This is what the message boxes look like now.

![messagebox1](https://cloud.githubusercontent.com/assets/6368949/8906671/aaa90e08-3470-11e5-9229-b9666a222aa0.png)
 * The messages are still largely unchanged from the first version that I just tossed in there to have something to work with. Probably the 'Recover' button explanation needs to be reworked a bit as it doesn't properly fit it's role.
 * The title text is wider than the table. Maybe some padding there?
 * Should I use normal text for the message line instead of header style?
 * Using the word 'Exit' instead of the default 'Abort' has one down side. The standard buttons are translated and overriding the name gives us one more string to translate.
 * @Umcaruje What would be good icons here? I'm thinking maybe something with an exclamation mark for ignore and then maybe the 'door with arrow' thing for exit. 

![closemsgrecovery](https://cloud.githubusercontent.com/assets/6368949/8906679/c0482abe-3470-11e5-982a-0fd204be6184.png)

![closemsgchanged](https://cloud.githubusercontent.com/assets/6368949/8906680/c4c67c26-3470-11e5-93c9-8a8906a12228.png)

These two works as they are now I think. The original 'Project not saved' message is still used when it applies. In case the work is both recovered and changed, you get the 'Recovered project not saved' message.

The debug messages are all coming out. I was thinking of maybe leaving some of them like the ones that indicate which session it is but that would probably just be confusing. Especially the 'Limited session' one as limited isn't used anywhere else but as an internal term. The user has just pressed a button and it should be obvious what happens.

```
Recover session...
Limited session...
```

8tab commented:
I think you should replace sentence: "Please don't run multiple instances of LMMS when you do this" with "Please close other instances of LMMS when you do this". Moreover you can detect if there is other instance of LMMS running and disallow recovering.

zonkmachine commented:
But it's OK to recover a file with other instances running.

8tab commented:
Ok, so do we really need this message then?

zonkmachine commented:
Well. You can open a file with recover and then open other sessions with ignore and that's all fine. If you open LMMS normally in one session and then go for a coffee, comes back and forgets about your session already running, now you are presented with a recover message box if you start a new LMMS window. That could be confusing. Now detecting other running instances of LMMS sounds lika a very good idea, but I'd have to look that up as I've never done that. Not in a way that would be approved around here anyway ;)
I think we could cover this with a well phrased message instead.

Umcaruje commented:
>Should I use normal text for the message line instead of header style?

I think that would look nicer, it looks a bit too large and bold to me rn.

>@Umcaruje What would be good icons here? I'm thinking maybe something with an exclamation mark for ignore and then maybe the 'door with arrow' thing for exit.

Maybe a crossed circle for ignore (:no_entry_sign:)? 'door with arrow' thing sounds good for exit.

zonkmachine commented:
![recovernoparagraph](https://cloud.githubusercontent.com/assets/6368949/8912819/75c8e29c-3495-11e5-958a-4e64cbbf8947.png)

Normal letters in message and padding on the side. I also tried to give a little padding in between the message and the table but so far no success.

zonkmachine commented:
Here's info on the subset of HTML 4 used in Qt 4.8
http://doc.qt.io/qt-4.8/richtext-html-subset.html

zonkmachine commented:
> Maybe a crossed circle for ignore (:no_entry_sign:)? 'door with arrow' thing sounds good for exit.

![msgboxexit](https://cloud.githubusercontent.com/assets/6368949/8947045/7eabaec4-3596-11e5-8866-0784fee89402.png)

I tried the existing exit.png (from Quit in the menu) I'm a bit sceptic to the 'no entry' take on ignore but I can't think of any other icons that would fit either.

zonkmachine commented:
@tresf 
I can set custom icons with ```button->setIcon( embed::getIconPixmap( "icon" );``` but how do I get the icons into the 'embed' function?
Right now I can't add new icons, only use the ones that are already in the system. I have added icons before but this time it doesn't seem to work to just drop them in the default themes folder.

tresf commented:
@zonkmachine I'm not sure we hard code them anywhere... Which folder are you adding them to?  If you overwrite a known-good pixmap, is it reflected in the software?  You might be accidentally pulling from the wrong data directory.

zonkmachine commented:
A ghost in the machine... It eventually just turned up. :-/ -?

![noentrymsgbox](https://cloud.githubusercontent.com/assets/6368949/9029798/8fecda9e-3999-11e5-837b-3b0363495378.png)

There are better looking no entry signs out there, I know.

tresf commented:
:+1: 

> There are better looking no entry signs out there, I know.

Easy to replace.  The critics can make a new one. :)

zonkmachine commented:
Squash it down to one commit? ...

tresf commented:
I would say yes. :+1:

zonkmachine commented:
:tractor:

zonkmachine commented:
One last debug message removed.

zonkmachine commented:
@tresf Any more comments on this one? I suggest we merge this now.

tresf commented:
The only other feedback that I have which I've just noticed now is the amount of functions this adds to MainWindow.cpp and the amount of boolean member variables that get added.  It seems like a lot of functions and variables for what appears to be a state variable storing a few mutually exclusive flags.

zonkmachine commented:
> and the amount of boolean member variables that get added.

I commented on this one third up from here:
https://github.com/LMMS/lmms/pull/2176#discussion_r35378740

tresf commented:
> I commented on this one third up from here [...]

Ah yes...

tresf commented:
@zonkmachine this has some merge conflicts that must be resolved prior to merging.  Can we also add the state variable as well via https://github.com/LMMS/lmms/pull/2176#discussion_r35378740?

Marking for 1.2 milestone.

zonkmachine commented:
On the case.

zonkmachine commented:
1 - I'll try and update this but I've never fixed a merge conflict before.

2 -  ```Can we also add the state variable as well via #2176 (comment)?```

I won't have time to fix that right now. Id rather sneak back later and fix it.

tresf commented:
> 1 - I'll try and update this but I've never fixed a merge conflict before.

It usually is a matter of doing a rebase and finding the conflicting lines and making them what they **should** be.

For example, if you changed:

```c++
#Foo.cpp
QString foo = "bar";
```

and in the mean time someone else changed:

```
#Foo.cpp
QString foo = "bar1";
```

Then you may have something like this in `Foo.cpp`

```diff
<<<<<<<
- QString foo = "bar";
=======
+ QString foo = "bar1";
>>>>>>>
```

And then you have to decide which line to keep, since the commit history will have to remain.  In most cases, the new value is the one you want, although if a function or class is refactored, it can be much more complicated to merge a conflict.

tresf commented:
> I won't have time to fix [the state variable] right now. Id rather sneak back later and fix it.

@Wallacoloo and @michaelgregorius can chime in, but I'd rather not expose 8 new functions to `MainWindow.h` for a single dialog, that seems like overkill.

zonkmachine commented:
OK. I'll try a rebase and will have a look at the functions in the weekend.

Wallacoloo commented on include/MainWindow.h:
```diff
@@ -81,6 +81,42 @@ class MainWindow : public QMainWindow
+	bool isRecoverSession()
+	{
+		return m_recoverSession;
+	}
```
I believe most of these functions can be kept private.

zonkmachine replied:
I'll look into it.

Wallacoloo commented:
@tresf One could make an argument that these functions would be better suited for `GuiApplication.h` since they mostly act in the background rather than visibly affecting anything inside the main window. By this same logic, most of the autosave functionality should be moved there as well. If we were designing things from scratch, this is what I would propose, but since MainWindow seems to already be taking on the responsibility of all 'etc' tasks, it almost seems best to keep it here for consistency.

Perhaps one improvement though is that the functions like `setLimitedSession()`, `resetLimitedSession()` could be reduced to one `setLimitedSession(bool isLimitedSession)` function.

zonkmachine commented:
> One could make an argument that these functions would be better suited for GuiApplication.h

I'm going to do the simplest possible adaptation to get this committed so this is probably not for me right now.

zonkmachine commented:
> For example, if you changed: ...

A no go! I don't know if it's because there's too many changes but I couldn't update it so I'm going to apply the changes "manually".

zonkmachine commented:
> @zonkmachine this has some merge conflicts that must be resolved prior to merging. Can we also add the state variable as well via #2176 (comment)?

Merge conflicts fixed. I'm looking into the suggested improvements now.

zonkmachine commented:
OK. I swapped the bools for a sessioin state int/enum. The autosave timer seem to be acting up right now though.

zonkmachine commented:
> Edit2! The "enum SessionState" commit isn't working (crashes).

I haven't been able to repeat the crashes I saw. It seem to work as intended. I disabled 'open last project' on limited session as the last project in this case probably was the crashed one and should be opened with recover.

zonkmachine commented:
@tresf @Wallacoloo 
I've updated the pull request and ditched the session state bools. I haven't looked into public functions that can be made/kept private. I call most of them both from main.cpp and MainWindow.cpp

tresf commented:
Thanks.  Much cleaner with the enum state.  I only had one comment to make.  Interested in hearing from @Wallacoloo.

zonkmachine commented:
Squashed and done. I've updated the Issue/OP.
The tests I devised in https://github.com/LMMS/lmms/pull/2176#issuecomment-125035898 is still passed.

zonkmachine commented:
Just a last minute tweak. I changed the name of function recoverSessionCleanup() to sessionCleanup() .

zonkmachine commented:
@tresf I'm think I'm done here. I've found other things related to the auto save function to work on but I think it's probably better to get this committed and start from a fresh pull request.

tresf commented:
@zonkmachine much cleaner, thanks.

I'll give some time for final comments prior to merging.

zonkmachine commented:
@tresf @Wallacoloo @Umcaruje 
I want to move on here and I don't think there is anything else to do on this PR.
One last change though. I reverted one line as I had accidentally changed how the .bak/backup file worked. Not touching that file no more.

tresf commented:
@zonkmachine sorry for the delay.  Merging. :+1:

musikBear commented:
RC3 -An unwanted change has sneaked in. I have several times had lmms 'stagger' during song-editor replay, and if i emediately after such a stagger, look at the recover.mmp, is has that exact timestamp. This mean that the protection against making a recover-file during replay, no longer works in RC3
My chosen recover interval is 10 mins, in *Settings*.
The unwanted save does not happens regulary, or predictable

zonkmachine commented:
Thanks, looking into it.

zonkmachine commented:
@musikBear What does your setting looks like?

zonkmachine commented:
> This mean that the protection against making a recover-file during replay, no longer works in RC3

In RC2 we introduced an option to allow auto-save while playing. The controls were incorrect however.
In RC3 tresf fixed this and it's working correct for me. Could it be that you've simply got the wrong settings? 
![recover](https://user-images.githubusercontent.com/6368949/28376757-3ab1e46a-6cab-11e7-9d92-eb3a5bc08ab8.png)

musikBear commented:
> Could it be that you've simply got the wrong settings?

🤡 o ..m ..g 🤡 + 😊 
Somehow i must have chosen both
Or..
@zonkmachine 
Could it be that *if* autosave + interval already are chosen in the **previous version (RC2)** that **both** auto-save-options actually will be **default** *true* in RC3?
I doubt i made any setting changes after i upgraded from RC2 to RC3.
-and come to think of it, thats actually a bad test-protocol, because then issues like this one will be obscured!

zonkmachine commented:
> Could it be that if autosave + interval already are chosen in the previous version (RC2) that both auto-save-options actually will be default true in RC3?

No. But the default is that both options are on. In your case you need to set this off every time you have a brand new .lmmsrc.xml

musikBear commented:
@zonkmachine I dont like that descission. That feature should not be on **in replay** as default. 
Default should be in idle-mode, and i think default time-setting could be  ..4 mins. Looking at the things that gives new users 'problems', has forced me to understand: 
* Nobody likes to explore a new program
* Absolutely Nobody care to read manuals
:0)
So having default autosave in replay, will only lead to Post in both forum and hub of the nature:
*"My LMMS cant replay my song right, all the time"*
🙊

zonkmachine commented:
@musikBear https://github.com/LMMS/lmms/pull/3722
